File: backend/app/models/itinerary.py
import uuid
from enum import Enum as PyEnum

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy import Enum as SAEnum
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ActivityStatus(str, PyEnum):
    planned = "planned"
    done = "done"
    skipped = "skipped"


class ItineraryDay(Base):
    __tablename__ = "itinerary_days"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    trip_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("trips.id", ondelete="CASCADE"), nullable=False)
    date: Mapped[str] = mapped_column(String, nullable=False)
    day_number: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    activities: Mapped[list["Activity"]] = relationship(
        "Activity", back_populates="day", cascade="all, delete-orphan", order_by="Activity.order_index"
    )


class Activity(Base):
    __tablename__ = "activities"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    day_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("itinerary_days.id", ondelete="CASCADE"), nullable=False)
    trip_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("trips.id", ondelete="CASCADE"), nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    types: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)
    start_time: Mapped[str | None] = mapped_column(String, nullable=True)
    end_time: Mapped[str | None] = mapped_column(String, nullable=True)
    map_link: Mapped[str | None] = mapped_column(String, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    order_index: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    status: Mapped[ActivityStatus] = mapped_column(SAEnum(ActivityStatus, name="activitystatus"), nullable=False, default=ActivityStatus.planned)
    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    day: Mapped["ItineraryDay"] = relationship("ItineraryDay", back_populates="activities")

File: frontend/types/itinerary.ts
export type ActivityType = string;

export type ActivityStatus = "planned" | "done" | "skipped";

export interface Activity {
  id: string;
  day_id: string;
  trip_id: string;
  title: string;
  types: ActivityType[];
  start_time: string | null;
  end_time: string | null;
  map_link: string | null;
  notes: string | null;
  order_index: number;
  status: ActivityStatus;
  created_at: string;
}

export interface ItineraryDay {
  id: string;
  trip_id: string;
  date: string;
  day_number: number;
  activities: Activity[];
  created_at: string;
}

export interface ActivityCreate {
  title: string;
  types?: ActivityType[];
  start_time?: string;
  end_time?: string;
  map_link?: string;
  notes?: string;
  status?: ActivityStatus;
}

export interface ActivityUpdate {
  title?: string;
  types?: ActivityType[];
  start_time?: string | null;
  end_time?: string | null;
  map_link?: string | null;
  notes?: string | null;
  status?: ActivityStatus;
}
